modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java

```diff
@@ -87,21 +95,42 @@ public static Object apply(Object fieldValue, boolean allowDuplicateKeys) {
         }
     }
 
-    public static void apply(Map<String, Object> ctx, String fieldName, boolean allowDuplicateKeys) {
+    public static void apply(Map<String, Object> ctx, String fieldName, boolean allowDuplicateKeys, boolean addToRootRecursiveMerge) {
         Object value = apply(ctx.get(fieldName), allowDuplicateKeys);
         if (value instanceof Map) {
             @SuppressWarnings("unchecked")
                 Map<String, Object> map = (Map<String, Object>) value;
-            ctx.putAll(map);
+            if (addToRootRecursiveMerge) {
```

Comment:
Maybe pass down the `MergeStrategy` enum here? Instead of translating it to a boolean and then pass that down.